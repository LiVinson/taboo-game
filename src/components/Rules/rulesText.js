const rules = [
	{
		title: 'Setup',
		text: [
			'Taboo is a 4 -10 player team game that you can play when you are together with friends or virtually using platforms like Google Meets, Zoom, Skype, or even a group phone call.',
			'Once a new game is created, join using the gamecode, select a team, and start the fun.',
		],
	},
	{
		title: 'Clue Giver',
		text: [
			'Each round, a clue giver is chosen. The giver sees a card with one word at the top of the card and a list of words below.',
			'The goal is to get your team to the word at the top by giving them a description of the word without saying anything taboo.',
		],
	},
	{
		title: 'Clue Giver (cont.)',
		text: [
			"The giver can not say any part of the words on the card. For example, if the word to guess is 'microwave' the clue giver cannot say micro, microphone, waving, etc.",
			"The clue giver cannot use 'rhymes with' or 'sounds like' clues. Props or pointing at objects are not allowed.",
		],
	},
	{
		title: 'Watcher',
		text: [
			'A watcher is chosen from the other team and can also see the taboo card. If the giver says or does anything taboo, the watcher selects the Buzzer button to force the giver to move to the next card in the pile.',
			"The watcher's teammates should not guess or help the clue giver's team in any way.",
		],
	},
	{
		title: 'Scoring',
		text: [
			'After each round, the watcher reviews the cards and confirms each is the right pile: correct, skipped, or discarded.',
			"Correct cards are worth one point for the giving team. Skipped cards are worth 0, Â½, or 1 point for the watcher's team, and discarded cards are worth 0.",
			'The game ends when every player has been the giver a set number of times, or all cards have been played. The team with the most points wins!',
		],
	},
]

export default rules
