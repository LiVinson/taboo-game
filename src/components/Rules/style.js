import styled from 'styled-components'

export const RulesTitle = styled.h5`
	font-family: ${(props) => props.theme.text.display};
	font-size: 2rem;
	font-weight: 500;
	text-align: center;
	color: ${(props) => props.theme.color.grayDark2};

	/* Faint gray line after section title*/
	&::after {
		display: block;
		content: '';
		width: 80%;
		height: 1.5px;
		background-color: ${(props) => props.theme.color.lightGray2};
		margin: 0.8rem auto 1rem auto;
	}
`

export const RulesText = styled.p`
	font-size: 1.6rem;
	text-align: left;
	padding: 0 0.5rem;
	margin-bottom: 1rem;
	color: ${(props) => props.theme.color.grayDark2};
`
