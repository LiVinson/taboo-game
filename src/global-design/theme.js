const theme = {
  color: {
    primary: "#1982C4", //light blue
    accent1: "#8AC926", //lime green
    accent2: "#ee5b2c", //orange
    accent3: "#a23b72", //purple
    primaryLight: "#58b1e9", //light-blue
    white: "#fff",
    offWhite: "#f7f9f9",
    lightGray1: "#f1f1f1",
    lightGray2: "#ebebeb",
    grayDark1: "#757575",
    grayDark2: "#333333"
  },
  text: {
    display: "'Oswald', sans-serif",
  },
  fontSize: {
    xsmall: "1.8rem",
    small: "2.2rem",
    medium: "2.7rem",
    large: "4rem",
    xlarge: "7rem",
  },
}


export default theme
