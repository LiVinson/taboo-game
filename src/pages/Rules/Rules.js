import React from "react"
import TabooCard from "../../components/TabooCard"
import { Redirect } from "react-router-dom"
import rules from "./rulesText"

export default function Rules({ match }) {
  const { topic } = match.params
  console.log(topic)
  return (
    <TabooCard
         tabooWord="How to Play" list={null}
         buttons={true}
    >
      <RulesText topic={topic} />
    </TabooCard>
  )
}

function RulesText({topic}) {
    console.log(topic)

    const rulesInfo = rules[parseInt(topic)]

    if (rulesInfo) {
        return (
            <React.Fragment>
                <h5>{rulesInfo.title}</h5>
                {rulesInfo.text.map((paragraph, index) => <p key={index}>{paragraph}</p>)}
            </React.Fragment>
        )
        
    } else {
        return <Redirect to="/404" />

    }
}


