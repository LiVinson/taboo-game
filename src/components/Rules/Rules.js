import React from 'react'
import { Redirect } from 'react-router-dom'
import PropTypes from 'prop-types'
import { ButtonTabooCard } from 'components/shared/TabooCard'
import rules from './rulesText'
import { RulesTitle, RulesText } from './style'

/*
  Can be displayed in 2 ways: on it's own route or within the play route.
  If on home route closing instructions goes back to /home. If on play route,
  calls a callback provided from GameInfo that will toggle Rules closed.
*/
export default class Rules extends React.Component {
	constructor(props) {
		super(props)
		this.state = {
			currentRule: 0,
		}
	}

	//When user clicks Back. If already showing first rule, either call toggleGameInfo (during game play)
	//or return to home route. Otherwise set state to previous rule
	displayPreviousRule = (event) => {
		const { currentRule } = this.state
		if (currentRule > 0) {
			this.setState((state) => {
				return {
					currentRule: state.currentRule - 1,
				}
			})
		} else {
			if (this.props.toggleGameInfo) {
				this.props.toggleGameInfo(event)
			} else {
				this.props.history.push('/home')
			}
		}
	}

	//When user clicks Next. If already showing last rule, either call toggleGameInfo (during game play)
	//or return to home route. Otherwise set state to next rule
	displayNextRule = (event) => {
		const { currentRule } = this.state
		if (currentRule < rules.length - 1) {
			this.setState((state) => {
				return {
					currentRule: state.currentRule + 1,
				}
			})
		} else {
			if (this.props.toggleGameInfo) {
				this.props.toggleGameInfo(event)
			} else {
				this.props.history.push('/home')
			}
		}
	}

	render() {
		const { currentRule } = this.state
		const buttonInfo = [
			{
				text: currentRule > 0 ? 'Back' : 'Close',
				onClick: this.displayPreviousRule,
				name: this.props.toggleGameInfo && 'showRules',
				type: 'button',
			},
			{
				text: currentRule < rules.length - 1 ? 'Next' : 'Close',
				onClick: this.displayNextRule,
				name: this.props.toggleGameInfo && 'showRules',
				type: 'button',
			},
		]

		return (
			<ButtonTabooCard tabooWord="How to Play" buttons={buttonInfo} type="home">
				<DisplayRulesText topicId={currentRule} />
			</ButtonTabooCard>
		)
	}
}

Rules.propTypes = {
	history: PropTypes.object,
	toggleGameInfo: PropTypes.func,
}

//Loops over imported array of rules and displays title and paragraphs for rule displayed
const DisplayRulesText = ({ topicId }) => {
	const rulesInfo = rules[topicId]
	if (rulesInfo) {
		return (
			<React.Fragment>
				<RulesTitle>{rulesInfo.title}</RulesTitle>
				{rulesInfo.text.map((paragraph, index) => (
					<RulesText key={index}>{paragraph}</RulesText>
				))}
			</React.Fragment>
		)
	} else {
		//Shouldn't be possible, but just in case.
		return <Redirect to="/404" />
	}
}

DisplayRulesText.propTypes = {
	topicId: PropTypes.number.isRequired,
}
