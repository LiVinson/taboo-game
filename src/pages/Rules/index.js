import Rules from "./Rules"

export default Rules