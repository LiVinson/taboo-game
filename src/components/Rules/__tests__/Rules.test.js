import React from 'react'
import renderer from 'react-test-renderer'
import { shallow, mount } from 'enzyme'
import 'jest-styled-components'
import { ThemeProvider } from 'styled-components'
import theme from '../../../global-design/theme'
import Rules from '../Rules'
import rulesText from '../rulesText'

test('Rules renders without crashing', () => {
	const props = {
		history: {
			push: jest.fn(),
		},
	}

	shallow(<Rules {...props} />)
})

test('Rules render correctly', () => {
	const props = {
		history: {
			push: jest.fn(),
		},
	}

	const wrapper = renderer
		.create(
			<ThemeProvider theme={theme}>
				<Rules {...props} />
			</ThemeProvider>
		)
		.toJSON()

	expect(wrapper).toMatchSnapshot()
})

test('currentRule state decrement/increments on click of Back/Next button', () => {
	const props = {
		history: {
			push: jest.fn(),
		},
	}

	const wrapper = mount(
		<ThemeProvider theme={theme}>
			<Rules {...props} />
		</ThemeProvider>
	)
	wrapper.find('button').at(1).simulate('click')
	expect(wrapper.find(Rules).state('currentRule')).toBe(1)

	wrapper.find('button').at(0).simulate('click')
	expect(wrapper.find(Rules).state('currentRule')).toBe(0)
})

test("Rules navigates to homepage if 'back' is selected and first rule is already displayed", () => {
	const props = {
		history: {
			push: jest.fn(),
		},
	}
	const wrapper = mount(
		<ThemeProvider theme={theme}>
			<Rules {...props} />
		</ThemeProvider>
	)
	wrapper.find('button').at(0).simulate('click')

	//Check that the history method was called
	expect(props.history.push.mock.calls[0].length).toBe(1)
	//Check the argument the push method was called with (equals the new url)
	expect(props.history.push.mock.calls[0][0]).toBe('/home')
})

test('Navigates to /home once all rules clicked through', () => {
	const props = {
		history: {
			push: jest.fn(),
		},
	}
	const wrapper = mount(
		<ThemeProvider theme={theme}>
			<Rules {...props} />
		</ThemeProvider>
	)

  //Click next for each rules in rules array
	rulesText.forEach((rule) => {
		wrapper.find('button').at(1).simulate('click')
	})
	//Check that the history method was called
	expect(props.history.push.mock.calls[0].length).toBe(1)
	//Check the argument the function was called with
	expect(props.history.push.mock.calls[0][0]).toBe('/home')
})
