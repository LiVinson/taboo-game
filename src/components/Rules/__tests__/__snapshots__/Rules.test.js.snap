// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Rules render correctly 1`] = `
.c0 {
  width: 30rem;
  margin: 0 auto;
  padding: 2rem;
  border-radius: 2px;
  position: relative;
  background-color: currentColor;
  color: #1982C4;
}

.c0:before {
  z-index: -1;
  position: absolute;
  content: '';
  bottom: 1.2rem;
  left: 0.5rem;
  width: 50%;
  top: 80%;
  max-width: 300px;
  background: #777;
  box-shadow: 0 1.5rem 1rem rgba(0,0,0,0.4);
  -webkit-transform: rotate(-3deg);
  -ms-transform: rotate(-3deg);
  transform: rotate(-3deg);
}

.c1 {
  background-color: #f7f9f9;
  padding: 1.2rem;
  text-align: center;
  border-radius: 2px;
  font-family: 'Oswald',sans-serif;
  font-size: 4rem;
  color: inherit;
  font-weight: 400;
  text-transform: uppercase;
}

.c2 {
  background-color: #f7f9f9;
  padding: 1.5rem;
  text-align: center;
  border-radius: 2px;
  font-size: 4rem;
  margin-top: 1.5rem;
}

.c6 {
  width: 47%;
  padding: 0.6rem 0rem;
  border: none;
  border-radius: 2px;
  background-color: #f7f9f9;
  font-size: 2.3rem;
  font-family: 'Oswald',sans-serif;
  text-transform: uppercase;
  color: #1982C4;
}

.c6:hover {
  -webkit-transform: translateY(-2px);
  -ms-transform: translateY(-2px);
  transform: translateY(-2px);
}

.c6:active {
  -webkit-filter: brightness(0.8);
  filter: brightness(0.8);
  -webkit-transform: translateY(2px);
  -ms-transform: translateY(2px);
  transform: translateY(2px);
}

.c6:focus {
  border: none;
  outline: solid 2px #8AC926;
}

.c6:disabled {
  -webkit-filter: brightness(0.6);
  filter: brightness(0.6);
  cursor: not-allowed;
}

.c7 {
  width: 47%;
  padding: 0.6rem 0rem;
  border: none;
  border-radius: 2px;
  background-color: #f7f9f9;
  font-size: 2.3rem;
  font-family: 'Oswald',sans-serif;
  text-transform: uppercase;
  color: #1982C4;
  color: #f7f9f9;
  background-color: #8AC926;
}

.c7:hover {
  -webkit-transform: translateY(-2px);
  -ms-transform: translateY(-2px);
  transform: translateY(-2px);
}

.c7:active {
  -webkit-filter: brightness(0.8);
  filter: brightness(0.8);
  -webkit-transform: translateY(2px);
  -ms-transform: translateY(2px);
  transform: translateY(2px);
}

.c7:focus {
  border: none;
  outline: solid 2px #8AC926;
}

.c7:disabled {
  -webkit-filter: brightness(0.6);
  filter: brightness(0.6);
  cursor: not-allowed;
}

.c7:focus {
  outline: solid 2px #f7f9f9;
}

.c5 {
  margin-top: 1.5rem;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-box-pack: justify;
  -webkit-justify-content: space-between;
  -ms-flex-pack: justify;
  justify-content: space-between;
  -webkit-flex-wrap: wrap;
  -ms-flex-wrap: wrap;
  flex-wrap: wrap;
}

.c3 {
  font-family: 'Oswald',sans-serif;
  font-size: 2rem;
  font-weight: 500;
  text-align: center;
  color: #333333;
}

.c3::after {
  display: block;
  content: '';
  width: 80%;
  height: 1.5px;
  background-color: #ebebeb;
  margin: 0.8rem auto 1rem auto;
}

.c4 {
  font-size: 1.6rem;
  text-align: left;
  padding: 0 0.5rem;
  margin-bottom: 1rem;
  color: #333333;
}

<div
  className="c0"
>
  <div
    className="c1"
  >
    How to Play
  </div>
  <div
    className="c2"
  >
    <h5
      className="c3"
    >
      Setup
    </h5>
    <p
      className="c4"
    >
      Taboo is a 4 -10 player team game that you can play when you are together with friends or virtually using platforms like Google Meets, Zoom, Skype, or even a group phone call.
    </p>
    <p
      className="c4"
    >
      Once a new game is created, join using the gamecode, select a team, and start the fun.
    </p>
  </div>
  <div
    className="c5"
  >
    <button
      className="c6"
      onClick={[Function]}
      type="button"
    >
      Close
    </button>
    <button
      className="c7"
      onClick={[Function]}
      type="button"
    >
      Next
    </button>
  </div>
</div>
`;
